exporter/awsemfexporter/metadata/metadata.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadata
+
+import (
+	"bufio"
+	"io"
+	"os"
+
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+)
+
+const (
+	containerIDLength = 64
+	defaultCGroupPath = "/proc/self/cgroup"
+)
+
+type Metadata interface {
+	GetCollectorIdentifier() (string, error)
+}
+
+type metadata struct {
+	ec2    *ec2metadata.EC2Metadata
+	docker *DockerHelper
+	hostID string
+}
+
+func NewMetadata(s *session.Session, cGroupPath string) Metadata {
+	if cGroupPath == "" {
+		cGroupPath = defaultCGroupPath
+	}
+	return &metadata{
+		ec2: ec2metadata.New(s),
+		docker: &DockerHelper{
+			cGroupPath: cGroupPath,
+		},
+	}
+}
+
+func (m *metadata) getEC2InstanceID() (string, error) {
+	instance, err := m.ec2.GetInstanceIdentityDocument()
+	if err != nil {
+		return "", err
+	}
+	return instance.InstanceID, nil
+}
+
+func (m *metadata) getContainerID() (string, error) {
```

Comment:
Actually, I don't think container ID is a global unique value - the log stream needs to be globally unique within the log group so is container ID ok? A customer can configure the same log group for all services, even across different container clusters if they want to I think.

So while chance of collision of container ID between different clusters / hosts is low due to the size of the string, and that it's not common to share a log group so widely, I think it's not a correct mapping for this use case of guaranteeing unique stream names. UUID? :)

Reply:
This is a valid concern. Thanks. In this case, I think it's more safe to use UUID. :)